cloudinit/net/dhcp.py

```diff
@@ -503,62 +504,226 @@ def get_latest_lease(lease_dir, lease_file_regex):
+    def get_key_from_latest_lease(self, distro, key: str):
+        """Get a key from the latest lease from distro-managed dhclient
 
-        @param lease_file: Name of a file to be parsed
-        @return: An address if found, or None
+        Doesn't consider the ephemeral timeframe lease.
+
+        @param lease_dir: distro-specific lease to check
+        @param lease_file_regex: distro-specific regex to match lease name
+        @return: The most recent lease file, or None
         """
-        latest_address = None
-        with suppress(FileNotFoundError), open(lease_file, "r") as file:
-            for line in file:
-                if "dhcp-server-identifier" in line:
-                    words = line.strip(" ;\r\n").split(" ")
-                    if len(words) > 2:
-                        dhcptok = words[2]
-                        LOG.debug("Found DHCP identifier %s", dhcptok)
-                        latest_address = dhcptok
-        return latest_address
-
-
-class Dhcpcd:
+        lease_file = self.get_newest_lease_file_from_distro(distro)
+        if lease_file:
+            content: str
+            content = util.load_file(lease_file)  # pyright: ignore
+            if content:
+                for lease in reversed(self.parse_leases(content)):
+                    server = lease.get(key)
+                    if server:
+                        return server
+
+
+class Dhcpcd(DhcpClient):
     client_name = "dhcpcd"
 
-    def __init__(self):
-        raise NoDHCPLeaseMissingDhclientError("Dhcpcd not yet implemented")
+    def dhcp_discovery(
+        self,
+        interface: str,
+        dhcp_log_func: Optional[Callable] = None,
+        distro=None,
+    ) -> Dict[str, Any]:
+        """Run dhcpcd on the interface without scripts/filesystem artifacts.
+
+        @param interface: Name of the network interface on which to send a
+            dhcp request
+        @param dhcp_log_func: A callable accepting the client output and
+            error streams.
+        @param distro: a distro object for network interface manipulation
+        @return: dict of lease options representing the most recent dhcp lease
+            parsed from the dhclient.lease file
+        """
+        LOG.debug("Performing a dhcp discovery on %s", interface)
+
+        # ISC dhclient needs the interface up to send initial discovery packets
+        # Generally dhclient relies on dhclient-script PREINIT action to bring
+        # the link up before attempting discovery. Since we are using
+        # -sf /bin/true, we need to do that "link up" ourselves first.
+        distro.net_ops.link_up(interface)
+
+        # TODO: disabling hooks means we need to get all of the files in
+        # /lib/dhcpcd/dhcpcd-hooks/ and pass each of those with the --nohook
+        # argument to dhcpcd
+        try:
+            out, err = subp.subp(
+                [
+                    self.client_name,
+                    "--oneshot",  # get lease then exit
```

Comment:
@holmanb I think we have a problem with --oneshot on dhcpd, it kills all dhcpcd processes upon success... making the next call to dumplease reject due to no dhcpcd services running:

root@ip-192-168-2-172:~# dhcpcd --oneshot --nobackground --ipv4only --waitip=4 --persistent --noarp --script=/bin/true ens5
dhcpcd-10.0.6 starting
DUID 00:01:00:01:2d:3a:fc:cc:02:f2:c2:1c:ca:1f
ens5: IAID c2:1c:ca:1f
ens5: rebinding lease of 192.168.2.172
ens5: leased 192.168.2.172 for 3600 seconds
ens5: adding route to 192.168.0.0/20
ens5: adding default route via 192.168.0.1
exiting due to oneshot
dhcpcd exited
root@ip-192-168-2-172:~# dhcpcd --dumplease ens5
dhcpcd is not running
root@ip-192-168-2-172:~# echo $?
1

Reply:
I ended up filling some bugs in upstream dhcpcd and a couple of potential fixes in PRs too. For now we'll have to just use the daemon and make sure that we clean up after ourselves by killing it when we've accomplished what we needed. This requires use of the `-P` flag to get the lease file name, since it isn't necessarily consistent and there is no way to set its location.